Module 2 Exercises.ipynb

```diff
@@ -1,415 +1,1093 @@
+    "id": "O76Qx6NQMWSr"
+   },
+   "source": [
+    "\n",
+    "2. Clean the education data (`estat_edat_lfse_04.tsv`) and put it into \"tidy\" format. Check that there is a unique non-missing key for each row. The resulting data should have five columns and 299,088 rows."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "GmUibPXdMWSs"
+   },
+   "source": [
+    "3. Do the same for the income data (`estat_nama_10r_2hhinc.tsv`) and the population data (`estat_demo_r_pjanaggr3.tsv`)."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 2,
+   "metadata": {
+    "colab": {
+     "base_uri": "https://localhost:8080/",
+     "height": 400
     },
+    "id": "9HYlkboyMWSr",
+    "outputId": "b4974597-71b2-4f90-ebfd-62b4488260e3"
+   },
+   "outputs": [],
+   "source": [
+    "# 1, 2, 3\n",
+    "\n",
+    "def load_data_frames(data_folder='data'):\n",
+    "    files = os.listdir(data_folder)\n",
+    "    data_frames = {}\n",
+    "    for file in files:\n",
+    "        file_path = os.path.join(data_folder, file)\n",
+    "        if file.endswith('.tsv'):\n",
+    "            df_name = file.replace('.tsv', '')\n",
+    "            data_frames[df_name] = pd.read_csv(file_path, sep='\\t')\n",
+    "        elif file.endswith('.csv'):\n",
+    "            df_name = file.replace('.csv', '')\n",
+    "            data_frames[df_name] = pd.read_csv(file_path)\n",
+    "        elif file.endswith('.shp'):\n",
+    "            gdf_name = file.replace('.shp', '')\n",
+    "            data_frames[gdf_name] = gpd.read_file(file_path)\n",
+    "    globals().update(data_frames)\n",
+    "\n",
+    "def extract_number(value):\n",
+    "    match = re.search(r'(\\d+(\\.\\d+)?)', value)\n",
+    "    return float(match.group(1)) if match else None\n",
+    "\n",
+    "def process_dataframe(df, var_name):\n",
+    "    # Split the first column and set new column names\n",
+    "    first_cols = df.iloc[:, 0].str.split(',', expand=True)\n",
+    "    first_cols.columns = df.columns[0].split(',')\n",
+    "    df = pd.concat([first_cols, df.iloc[:, 1:]], axis=1)\n",
+    "    # Rename columns\n",
+    "    df.rename(columns={'geo\\\\TIME_PERIOD': 'geo'}, inplace=True)\n",
+    "    # Define new column names\n",
+    "    new_col_names = first_cols.columns.tolist()[:-1] + ['geo']\n",
+    "    # Reshape the DataFrame from wide to long format\n",
+    "    tidy_df = pd.melt(df, id_vars=new_col_names[:-1] + ['geo'], var_name='year', value_name='value')\n",
+    "    # Clean the 'value' column\n",
+    "    tidy_df[var_name] = tidy_df['value'].astype(str).apply(extract_number)\n",
+    "    tidy_df = tidy_df.drop(columns=[\"value\"])\n",
+    "    return tidy_df\n",
+    "\n",
+    "load_data_frames('data')\n",
+    "\n",
+    "estat_edat_lfse_04_tidy = process_dataframe(estat_edat_lfse_04, \"frac\")\n",
+    "estat_edat_lfse_04_tidy = estat_edat_lfse_04_tidy.loc[estat_edat_lfse_04_tidy['sex'] == \"T\"]\n",
```

Comment:
One perhaps more succinct way to right this (and the other subsetting steps below) is to use the `pandas.DataFrame.query`: 

```
estat_edat_lfse_04_tidy = estat_edat_lfse_04_tidy.query('sex == "T"')
```
You can also chain methods: 
```
estat_edat_lfse_04_tidy = (
    process_dataframe(estat_edat_lfse_04, "frac")
    .query('sex == "T"')
    .drop(columns=["freq", "sex", "unit"])
)
```